#1178 update validator generate subcommand to create encrypted BLS keystore

usmansaleem opened the pull request:
## PR Description
Update `teku validator generate` subcommand to generate encrypted keystore .

New cli options (as it appears in `teku validator generate --help`):
~~~
<snip>
-e, --encrypted-keystore-enabled=<true|false>
                            Create encrypted keystores for validator and
                              withdrawal keys. (Default: true)
<snip>
 Non-interactive password options for validator keystores:
      --validator-keystore-password-env=<ENV_VAR>
                            Read password from environment variable to encrypt
                              the validator keys
      --validator-keystore-password-file=<FILE>
                            Read password from the file to encrypt the
                              validator keys
Non-interactive password options for withdrawal keystores:
      --withdrawal-keystore-password-env=<ENV_VAR>
                            Read password from environment variable to encrypt
                              the withdrawal keys
      --withdrawal-keystore-password-file=<FILE>
                            Read password from the file to encrypt the
                              withdrawal keys
~~~

usmansaleem committed:
#1178 Generate encrypted keystore from validate generate subcommand

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
cli param labels

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
Merge remote-tracking branch 'upstream/master' into 1178_validator_generate_keystore

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
Merge remote-tracking branch 'upstream/master' into 1178_validator_generate_keystore

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
Add unit test case

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
Merge remote-tracking branch 'upstream/master' into 1178_validator_generate_keystore

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
update cli text

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
add line seperator in yaml line

Signed-off-by: Usman Saleem <usman@usmans.info>

macfarla commented on artemis/src/main/java/tech/pegasys/artemis/DepositCommand.java:
```diff
@@ -81,44 +99,144 @@ public void generate(
           Level.FATAL,
           "Failed to send deposit transaction: " + t.getClass() + ": " + t.getMessage());
-      System.exit(1); // Web3J creates a non-daemon thread we can't shut down. :(
+      exit(1);
```
has the web3j problem been solved?

usmansaleem replied:
I don't know about web3j. I moved the call in a method so that I can mock it in tests.

ajsutton replied:
Web3j remains problematic unfortunately.

jframe commented on artemis/src/main/java/tech/pegasys/artemis/DepositCommand.java:
```diff
@@ -81,44 +99,144 @@ public void generate(
+
+  private String generateRandomHexToken() {
+    final SecureRandom secureRandom = SecureRandomProvider.createSecureRandom();
+    byte[] token = new byte[32];
```
nit: final

usmansaleem replied:
This method has been removed.

jframe commented on artemis/src/main/java/tech/pegasys/artemis/DepositCommand.java:
```diff
@@ -81,44 +99,144 @@ public void generate(
+      final BLSKeyPair key, final Path outputDir, final String prefix, final int counter) {
+    try {
+      final KdfParam kdfParam = new SCryptParam(32, Bytes32.random());
+      final Cipher cipher = new Cipher(CipherFunction.AES_128_CTR, Bytes.random(16));
```
Think we should use SecureRandom for the random seed

usmansaleem replied:
tuweni Bytes random by default use `new SecureRandom`, but we should/can pass our own SecureRandom instance as well.

usmansaleem replied:
done.

jframe commented on artemis/src/main/java/tech/pegasys/artemis/DepositCommand.java:
```diff
@@ -81,44 +99,144 @@ public void generate(
+      Files.writeString(
+          keystoreDirectory.resolve(prefix + "_password_" + counter + ".txt"), randomPassword);
+    } catch (final IOException e) {
+      throw new UncheckedIOException(e);
```
Thinking we should log an error at this point so we know which keystore failed to be created

usmansaleem replied:
The methods has been extracted to a new class and they will be logging appropriate logging statements before erroring out.

jframe commented on artemis/src/main/java/tech/pegasys/artemis/DepositCommand.java:
```diff
@@ -81,44 +99,144 @@ public void generate(
+
     try (params) {
+      // generate BLS validator key and withdrawal key
+      final List<Pair<BLSKeyPair, BLSKeyPair>> validatorKeyPairs =
```
nit: instead the generic Pair a small custom class would help with readability so that you do something like validatorKeyPair.getValidatorKey() instead of getLeft for example

usmansaleem replied:
Its not used anymore.

jframe commented on artemis/src/main/java/tech/pegasys/artemis/DepositCommand.java:
```diff
@@ -81,44 +99,144 @@ public void generate(
+      final Path keyStoreOutputDirectory =
+          keystorePath == null || keystorePath.isBlank() ? Path.of(".") : Path.of(keystorePath);
+      final AtomicInteger counterSuffix = new AtomicInteger(0);
+      final List<SafeFuture<TransactionReceipt>> futures;
```
nit: transactionReceiptFutures?

jframe commented on artemis/src/test/java/tech/pegasys/artemis/DepositCommandTest.java:
```diff
@@ -0,0 +1,72 @@
+  }
+
+  private void assertKeyStoreAndPasswordExist(final Path parentDir, final int suffix) {
+    final Path keystore1File = parentDir.resolve("validator_keystore_" + suffix + ".json");
```
rename to validatorKeyStoreFile and similarly rename password1File to validatoryPasswordFile

usmansaleem replied:
This unit test has been refactored.

jframe commented on artemis/src/test/java/tech/pegasys/artemis/DepositCommandTest.java:
```diff
@@ -0,0 +1,72 @@
+  private void assertKeyStoreAndPasswordExist(final Path parentDir, final int suffix) {
+    final Path keystore1File = parentDir.resolve("validator_keystore_" + suffix + ".json");
+    final Path password1File = parentDir.resolve("validator_password_" + suffix + ".txt");
+    final Path keystore2File = parentDir.resolve("withdrawal_keystore_" + suffix + ".json");
```
rename withdrawalKeyStoreFile and similarly rename password2File to withdrawalPasswordFile

usmansaleem replied:
This unit test has been refactored.

ajsutton commented on artemis/src/main/java/tech/pegasys/artemis/DepositCommand.java:
```diff
@@ -81,44 +99,144 @@ public void generate(
+          String outputFile,
+      @Option(
+              names = {"--encrypt-keys", "-e"},
+              defaultValue = "false",
```
False doesn't seem like the right default here.  Writing unencrypted keys to disk is a really bad idea.

usmansaleem replied:
`--encrypt-keys` defaults to true now.

ajsutton commented on artemis/src/main/java/tech/pegasys/artemis/DepositCommand.java:
```diff
@@ -81,44 +99,144 @@ public void generate(
+          boolean encryptKeys,
+      @Option(
+              names = {"--keystore-output-dir", "-k"},
+              paramLabel = "<PATH>",
```
How does this relate to `--output-file`?  Seems kind of confusing to have two separate options to specify where to output things.  @jakehaugen any ideas on the right CLI options for this?

usmansaleem replied:
`--keystore-output-dir` is not used anymore. Instead I am now using `--output-path` to be treated either as directory or file depending upon we are using encrypted mode or unencrypted mode.

ajsutton commented on artemis/src/main/java/tech/pegasys/artemis/DepositCommand.java:
```diff
@@ -81,44 +99,144 @@ public void generate(
+  @VisibleForTesting
+  void exit(final int status) {
+    System.exit(status); // Web3J creates a non-daemon thread we can't shut down. :(
+  }
```
Rather than overriding this method in tests, it would be better to pass in a `Consumer<Integer> shutdownFunction` to the constructor (ie prefer delegation over inheritance).  You can pass a mock in tests and `System::exit` in production.

You will likely need two constructors for `DepositCommand` - a noarg one that passes `System::exit` and one that accepts the consumer that tests can use.

usmansaleem replied:
Implemented functional consumer interface.

ajsutton commented on artemis/src/main/java/tech/pegasys/artemis/DepositCommand.java:
```diff
@@ -81,44 +99,144 @@ public void generate(
+                        yamlStringBuilder
+                            .append(yamlFormattedString)
+                            .append(System.lineSeparator());
+                      }
```
Two issues here:
1. It's vital that we write the keys to output *before* sending the deposit transaction.  Otherwise if we fail somewhere we've just burned a lot of ETH and lost the keys forever.
2. Since the logic for writing keys has become more complex we should create a `KeyWriter` interface with 2 separate implementations - one that writes YAML to a `PrintStream` (used for all unencrypted output) and one that writes to a keystore for encrypted output.

usmansaleem replied:
Yes I hear you, thats why I wrote file in finally block that should have written keys up to sendTransaction fails. Using KeyWrite interface is better idea though.

ajsutton replied:
A finally block does not guarantee it happens first.  If the process or computer crashed the key would never be written and the funds would be lost.

usmansaleem replied:
We now have KeyWriter which is implemented by YamlKeysWriter and EncryptedKeyStoreWriter.

usmansaleem replied:
The approach to write keys has been changed so that they get persisted first before sendTransaction is called.

ajsutton commented on artemis/src/main/java/tech/pegasys/artemis/DepositCommand.java:
```diff
@@ -81,44 +99,144 @@ public void generate(
+
+      // save password
+      Files.writeString(
+          keystoreDirectory.resolve(prefix + "_password_" + counter + ".txt"), randomPassword);
```
We should have a way to specify the password to use for keystrokes.  No point writing an encrypted keystore if you write an unencrypted password for it as well.  It's probably enough to be able to specify a password for all signing keys and a (potentially different) password for all withdrawal keys.

usmansaleem replied:
that make sense. Let me figure out whether to read the password from stdin or from files or from cli option.

ajsutton replied:
The password should not be included on the command line.  Using `ps auxwww` any user on the same box can see command line arguments for any running program.  It also tends to get written to places like `.bash_history`

usmansaleem replied:
password input file it is ... unless we are open for "interactive shell" approach to read the input from user.

ajsutton replied:
I think an interactive mode would be an excellent option to support.  Need to check with POs as to priority and probably get Jake to think about the details.

usmansaleem replied:
👍

usmansaleem replied:
The interactive mode is now supported, along with reading values from environment variable as well as from a file.
~~~
--validator-password
--validator-password:env
--validator-password:file
~~~

ajsutton commented on artemis/src/main/java/tech/pegasys/artemis/DepositCommand.java:
```diff
@@ -81,44 +99,144 @@ public void generate(
+    final SecureRandom secureRandom = SecureRandomProvider.createSecureRandom();
+    byte[] token = new byte[32];
+    secureRandom.nextBytes(token);
+    return new BigInteger(1, token).toString(16); // hex encoding
```
It would be much better to use:
```suggestion
    return Bytes.wrap(token).toHexString()
```

usmansaleem replied:
This method has been removed.

ajsutton commented on artemis/src/test/java/tech/pegasys/artemis/DepositCommandTest.java:
```diff
@@ -0,0 +1,72 @@
+  private static final int VALIDATORS_COUNT = 2;
+  @Mock private DepositCommand.CommonParams commonParams;
+  @Mock private DepositTransactionSender depositTransactionSender;
+  @Spy private DepositCommand depositCommand = new DepositCommand();
```
Using a spy for the class under test is really unpleasant.

We also normally use the Mockito static methods instead of the `@Mock` annotations. Not for any particular reason, but it's pretty consistent so seems worth sticking with it.

usmansaleem replied:
The unit test code has been updated as per the suggestions.

ajsutton commented on artemis/src/test/java/tech/pegasys/artemis/DepositCommandTest.java:
```diff
@@ -0,0 +1,72 @@
+
+  @Test
+  void encryptedKeystoresAreCreated(@TempDir final Path tempDir) throws Exception {
+    doReturn(depositTransactionSender).when(commonParams).createTransactionSender();
```
This should be able to use the when format:
```suggestion
    when(commonParams.createTransactionSender()).thenReturn(depositTransactionSender);
```

That way we get type checking done by the compiler.

usmansaleem replied:
👍

ajsutton commented on artemis/src/test/java/tech/pegasys/artemis/DepositCommandTest.java:
```diff
@@ -0,0 +1,72 @@
+  @Test
+  void encryptedKeystoresAreCreated(@TempDir final Path tempDir) throws Exception {
+    doReturn(depositTransactionSender).when(commonParams).createTransactionSender();
+    doNothing().when(depositCommand).waitForTransactionReceipts(any());
```
The mock transaction sender should just return completed futures and you won't need to override this.

usmansaleem replied:
Thanks for the hint, I was having some trouble with it :). It's sorted out now.

ajsutton commented on artemis/src/test/java/tech/pegasys/artemis/DepositCommandTest.java:
```diff
@@ -0,0 +1,72 @@
+    Assertions.assertThat(tempDir.toFile().listFiles()).containsExactlyInAnyOrder(subDirectories);
+
+    for (int i = 0; i < subDirectories.length; i++) {
+      assertKeyStoreAndPasswordExist(subDirectories[i].toPath(), i + 1);
```
Is it worth checking we can actually decrypt the keystore?

usmansaleem replied:
The tests are now decrypting the keystore.

ajsutton commented on gradle/versions.gradle:
```diff
@@ -84,7 +84,8 @@ dependencyManagement {
 
     dependency 'org.mockito:mockito-core:3.1.0'
-    
+    dependency 'org.mockito:mockito-junit-jupiter:3.1.0'
```
We can avoid adding this dependency if we stay consistent with the `Mockito.mock` approach to creating mocks right?

usmansaleem replied:
reverted the dependency.

ajsutton reviewed (changes requested):
Heading in the right direction - the really critical thing is to ensure keys are written before we send transactions.

usmansaleem committed:
review suggestion - use functional interface to invoke System.exit

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
review suggestion - use static mock function

 -- removed mockito junit5 extensions
 -- updated mock senddeposittransaction to return completed future

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
Merge remote-tracking branch 'upstream/master' into 1178_validator_generate_keystore

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
wip

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
Merge remote-tracking branch 'upstream/master' into 1178_validator_generate_keystore

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
Use new KeysWriter in DepositCommand.
TODO: Update unit test case

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
Merge upstream branch

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
Interactive mode to read password. Also use arggroup

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
Merge remote-tracking branch 'upstream/master' into 1178_validator_generate_keystore

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
arg group not working, removed it temporarily

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
cleaning up interactive mode

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
fixing unit test

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
merge upstream

Signed-off-by: Usman Saleem <usman@usmans.info>

# Conflicts:
#	artemis/src/main/java/tech/pegasys/artemis/DepositCommand.java

usmansaleem committed:
merge conflicts status_log

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
fixing depositsender acceptance test

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
fixing depositsender acceptance test

Signed-off-by: Usman Saleem <usman@usmans.info>

ajsutton commented on acceptance-tests/src/test-support/java/tech/pegasys/artemis/test/acceptance/dsl/ArtemisDepositSender.java:
```diff
@@ -44,6 +47,11 @@ public String sendValidatorDeposits(final BesuNode eth1Node, final int numberOfV
+        .toString()
+        .lines()
+        .filter(s -> s.startsWith("- {"))
+        .collect(Collectors.joining(System.lineSeparator()));
```
This isn't the right approach.  If we're outputting to STDOUT, we need to make sure logging is either disabled or sent to STDERR.  Otherwise people who run this tool in real life will get corrupt output.

usmansaleem replied:
It's because of the limitation of "testcontainer" docker API we are using that captures whole of standard/error output. A more better approach, which I tested, is to mount the "keys" file with `--output-path` and read that file after running the command. However, in Windows environment this usually cause an issue because we would have "share" the host location via docker preference. In another project of ours, we used user's home folder and create a temp directory there which can be mounted to docker container, I wonder if we can use same approach. 

Last, but not least, is that I can remove the "info" logging that I introduced so that it only outputs the yaml keys on std out.

ajsutton replied:
It's not really docker related - if we're saying we'll output the content to STDOUT, we should *only* output the content to STDOUT but the logs are going there as well (they don't go to STDERR by default).  We can configure test containers to capture STDOUT and STDERR separately if need be.

I suspect just removing the log messages is probably the simplest approach though.

usmansaleem replied:
Removed additional logging for "yaml keys on standard out" mode.

ajsutton commented on artemis/src/main/java/tech/pegasys/artemis/DepositCommand.java:
```diff
@@ -80,45 +100,98 @@ public void generate(
+              arity = "1")
+          boolean encryptKeys,
+      @Option(
+              names = {"--validator-password:file"},
```
colon in the arg name is weird. I'd have expected `--validator-password-file`. Might be possible to do it with a single `--validator-password` option - if no arg use interactive prompt, if it starts with `env:` treat as an environment variable and otherwise treat as a file path.  Not entirely sure that's a good idea though...

@jakehaugen any thoughts?

usmansaleem replied:
I picked the :file :env approach from PicoCLI's sample code. We can modify them as per our needs. https://github.com/remkop/picocli/blob/master/picocli-examples/src/main/java/picocli/examples/interactive/PasswordDemo.java

ajsutton replied:
Yeah I have no idea why they thought that was normal...

usmansaleem replied:
You are suggesting to use `--stdin:validator-password`, `--env:validator-password`, `--file:validator-password` as an alternative? Or simply use three mutually exclusive parameters? i.e. `--validator-password-stdin`, `--validator-password-file`, `--validator-pasword-env`?

usmansaleem replied:
OR.... `--validator-password=env:VAL_PASS` approach?

ajsutton replied:
Two possible approaches:
1. Separate options.  So you'd have `--validator-password-file` and `--validator-password-env` if neither of those is specified, use the interactive prompt.

2. A single option that works out how to interpret the arg (feeling like this is going to be confusing, so leaning towards option 1).  So not specifying the option or `--validator-password` with no arg uses an interactive prompt.  `--validator-password env:SOME_VAR` would read from an environment var (the `env:` prefix is what tells us it's an env var) and `--validator-password <any-arg>` reads from a file (assuming <any-arg> doesn't start with env:).

ajsutton replied:
But you should really talk to Jake. That's why we have a UX person. :)

ajsutton commented on artemis/src/main/java/tech/pegasys/artemis/DepositCommand.java:
```diff
@@ -80,45 +100,98 @@ public void generate(
+              description = "Read password from environment variable to encrypt the validator keys")
+          String validatorPasswordEnv,
+      @Option(
+              names = {"--validator-password"},
```
It seems weird to have to specify `--validator-password` to get the prompt to provide the password. Makes me naturally want to specify the password on the command line which is insecure and doesn't actually work.

ajsutton commented on artemis/src/main/java/tech/pegasys/artemis/cli/deposit/EncryptedKeystorePasswordProvider.java:
```diff
@@ -0,0 +1,90 @@
+    this.errorPrefix = errorPrefix;
+  }
+
+  public String retrievePassword() {
```
Probably should check that only one of the three options for specifying the password have been provided.  Otherwise it will be very confusing which password actually gets used.

ajsutton commented on artemis/src/main/java/tech/pegasys/artemis/cli/deposit/EncryptedKeystorePasswordProvider.java:
```diff
@@ -0,0 +1,90 @@
+        String.format(
+            "Missing options: [--%1$s-password | --%1$s-password:env | --%1$s-password:file]",
+            errorPrefix);
+    throw new ParameterException(spec.commandLine(), errorMessage);
```
It's really annoying to get the missing options error message in small chunks.  ie I tried:
```
$ ./build/install/teku/bin/teku validator generate
Missing required options [--node-url=<URL>, --contract-address=<ADDRESS>, --private-key=<KEY>]

$ ./build/install/teku/bin/teku validator generate --node-url=http://localhost:8545 --contract-address=dddddddd --private-key=222222222
Missing options: [--validator-password | --validator-password:env | --validator-password:file]

$ ./build/install/teku/bin/teku validator generate --node-url=http://localhost:8545 --contract-address=dddddddd --private-key=222222222 --validator-password
Enter value for --validator-password (Password to encrypt validator keys):
Missing options: [--withdrawal-password | --withdrawal-password:env | --withdrawal-password:file]
```
So it took me four goes to actually be told all of what I had to provide.

usmansaleem replied:
I have introduced mutually exclusive ArgGroup so that picocli handles these errors automagically. I do have to extract out subcommands by converting methods to standalone classes to make ArgGroup work (I believe PicoCli has a bug where ArgGroup doesn't work if methods are used as subcommands). I have also changed the `--encrypt-keys` to `--encrypted-keystore-enabled` and we now run in interactive mode to obtain passwords if none of the non-interactive password options are specified. The ArgGroup also allows to provide sub-section heading. e.g. 

~~~
./teku validator generate --help

teku validator generate [OPTIONS]

Description:

Register validators by generating new keys and sending deposit transactions to
an Ethereum 1 node

Options:
  -a, --amount=<GWEI>       Deposit amount in Gwei (default: 32000000000)
  -c, --contract-address=<ADDRESS>
                            Address of the deposit contract
  -e, --encrypted-keystore-enabled=<true|false>
                            Create encrypted keystores for validator and
                              withdrawal keys. (Default: true)
  -h, --help                Show this help message and exit.
  -n, --number-of-validators=<NUMBER>
                            The number of validators to create keys for and
                              register
  -o, --output-path=<FILE|DIR>
                            Path to output file for unencrypted keys or output
                              directory for encrypted keystore files. If not
                              set, unencrypted keys will be written on standard
                              out and encrypted keystores will be created in
                              current directory
  -p, --private-key=<KEY>   Ethereum 1 private key to use to send transactions
  -u, --node-url=<URL>      JSON-RPC endpoint URL for the Ethereum 1 node to
                              send transactions via
  -V, --version             Print version information and exit.
Non-interactive password options for validator keystores:
      --validator-password-env=<ENV_VAR>
                            Read password from environment variable to encrypt
                              the validator keys
      --validator-password-file=<FILE>
                            Read password from the file to encrypt the
                              validator keys
Non-interactive password options for withdrawal keystores:
      --withdrawal-password-env=<ENV_VAR>
                            Read password from environment variable to encrypt
                              the withdrawal keys
      --withdrawal-password-file=<FILE>
                            Read password from the file to encrypt the
                              withdrawal keys

Teku is licensed under the Apache License 2.0
~~~

(@jakehaugen and @arash009  FYI)

usmansaleem replied:
Running generate command with required parameters only:

~~~
./teku validator generate --contract-address=dddddddddddddddddddddddddddddddddddddddd \
--number-of-validators=1 \
--private-key=8f2a55949038a9610f50fb23b5883af3b4ecb3c3bb792cbcefbd1542c692be63 \
--node-url=http://localhost:8545
Enter password for Validator Keystore:
Re-Enter password for Validator Keystore:
Enter password for Withdrawal Keystore:
Re-Enter password for Withdrawal Keystore:
22:18:24.268 [INFO ] - Generating Encrypted Keystores in .
~~~

arash009 replied:
To confirm, do you have to use the same method to supply the validator password and withdrawal password? e..g. can I supply validator password via env variable, and use interactive method for withdrawal password? (I dont think there really is a use case for that, but just checking what works)

usmansaleem replied:
@arash009  yes, both validator keystore and withdrawal keystore options are independent of each other, so validator keystore's password can come from environment variable for instance, while withdrawal's password can be read from file. However, due to mutually exclusive mode, you can either specify `--validator-password-env` or `--validator-password-file` but not both.

ajsutton commented on artemis/src/main/java/tech/pegasys/artemis/cli/deposit/EncryptedKeystoreWriter.java:
```diff
@@ -0,0 +1,95 @@
+  }
+
+  private Path createKeystoreDirectory() {
+    final Path keystoreDirectory = outputPath.resolve("validator_" + counter.incrementAndGet());
```
If I run the command twice, does this wind up overwriting the existing keys (losing my funds)?  Would it be better to use the public key as part of the directory name rather than the index - then it's guaranteed to be unique.

usmansaleem replied:
Yes, thats a good catch. I'll modify logic to use public key as folder name.

ajsutton reviewed (commented):
Looking good - just need to tidy up the UX a bit.  Probably worth talking to @jakehaugen about it.

usmansaleem committed:
Merge remote-tracking branch 'upstream/master' into 1178_validator_generate_keystore

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
fixing expected output in yamlkeyswritertest due to tuweni output in lower case

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
DepositCommand - reverting stdout logs

 -- reverted stdout logs when yaml keys are required to be printed on standard out
 -- fixed acceptance test case

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
Review suggestions - use validator bls public key as directory name

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
Converting subcommands methods to static classes to allow ArgGroups

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
Updating deposit subcommands and unit test cases

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
Merge remote-tracking branch 'upstream/master' into 1178_validator_generate_keystore

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
typo in exception msg

Signed-off-by: Usman Saleem <usman@usmans.info>

ajsutton reviewed (approved):
LGTM.

usmansaleem committed:
Merging upstream master

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
Review suggestion - update non-interactive option name to include keystore

Signed-off-by: Usman Saleem <usman@usmans.info>
